integration-tests/emulated/src/tests/peregrine/unpaid_execution.rs

```diff
@@ -0,0 +1,142 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2024 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+use frame_support::assert_ok;
+use xcm::{v3::WeightLimit, VersionedMultiLocation, VersionedXcm};
+use xcm_emulator::{
+	assert_expected_events, BodyId, BodyPart, Instruction::UnpaidExecution, Junction, Junctions, MultiLocation,
+	Outcome, Parachain, ParentThen, Plurality, RelayChain, TestExt, Xcm, X1,
+};
+
+use crate::mock::{
+	network::MockNetworkRococo,
+	para_chains::{peregrine, AssetHubRococo, AssetHubRococoPallet, Peregrine},
+	relay_chains::{Rococo, RococoPallet},
+};
+
+#[test]
+fn test_unpaid_execution_from_asset_hub_to_peregrine() {
+	MockNetworkRococo::reset();
+
+	let sudo_origin = <AssetHubRococo as Parachain>::RuntimeOrigin::root();
+	let parachain_destination: VersionedMultiLocation =
+		ParentThen(Junctions::X1(Junction::Parachain(peregrine::PARA_ID))).into();
+
+	let weight_limit = WeightLimit::Unlimited;
+	let check_origin = None;
+
+	let xcm = VersionedXcm::from(Xcm(vec![UnpaidExecution {
+		weight_limit,
+		check_origin,
+	}]));
+
+	//Send XCM message from Parachain
+	AssetHubRococo::execute_with(|| {
+		assert_ok!(<AssetHubRococo as AssetHubRococoPallet>::PolkadotXcm::send(
+			sudo_origin,
+			Box::new(parachain_destination),
+			Box::new(xcm)
+		));
+
+		type RuntimeEvent = <AssetHubRococo as Parachain>::RuntimeEvent;
+
+		assert_expected_events!(
+			AssetHubRococo,
+			vec![
+				RuntimeEvent::PolkadotXcm(pallet_xcm::Event::Sent { .. }) => {},
+			]
+		);
+	});
+
+	Peregrine::execute_with(|| {
+		type PeregrineRuntimeEvent = <Peregrine as Parachain>::RuntimeEvent;
+		assert_expected_events!(
+			Peregrine,
+			vec![
+				PeregrineRuntimeEvent::XcmpQueue(cumulus_pallet_xcmp_queue::Event::Fail {
+					error: xcm::v3::Error::Barrier,
+					..
+				}) => {},
+			]
+		);
+	});
+
+	// No event on Rococo. message is meant for Peregrine
+	Rococo::execute_with(|| {
+		assert_eq!(Rococo::events().len(), 0);
+	});
+}
+
+// TODO: Check why test is passing. Unpaid execution should work now.
+#[test]
+fn test_unpaid_execution_from_rococo_to_peregrine() {
+	MockNetworkRococo::reset();
+
+	let sudo_origin = <Rococo as RelayChain>::RuntimeOrigin::root();
```

Comment:
There appears to be an issue here. Since unpaid execution is now allowed, I need to establish a governance origin.

When I executed the test case with trace logs, I received this log message: `xcm::barriers: AllowUnpaidExecutionFrom origin: MultiLocation { parents: 1, interior: Here }, ...`

In the `WithComputedOrigin` step, the origin is the same:  `origin: xcm::barriers: WithComputedOrigin origin: MultiLocation { parents: 1, interior: Here }, ...`

@ntn-x2 Do you maybe know, how I can create a proper origin?

Reply:
Here are the logs: 

![image](https://github.com/KILTprotocol/kilt-node/assets/48685760/1bc4539e-5d70-454f-bdd8-7fb8f40e0090)

Reply:
Using the root location, gives parachain location, of course. This is equivalent to dispatching a call from the root track. To have the `ParentLegislative` origin, we would have to have a referendum on the relaychain. I think that's too much. Maybe we can go with the relay root origin directly, for now. I will check that. We need to investigate more how opengov origins get converted into `MultiLocation`s.

Reply:
Yeah the problem is the use of `.encode()`, which generates a vector that is prefixed by the length of the vector itself, which of course makes for an invalid asset DID. Currently we don't have a way to convert an asset DID to a raw vector, as far as I could see, so the test case needs to be constructed by hand, as I did here: https://github.com/KILTprotocol/kilt-node/pull/603/commits/b2cb68a105cf83da6b0e30ec31dbc6c5c597e91a.

Reply:
Using the root location, gives parachain location, of course. This is equivalent to dispatching a call from the root track. To have the `ParentLegislative` origin, we would have to have a referendum on the relaychain. Also, we would be testing with stuff that does not exist anymore, since all relaychains have migrated to opengov by now. Maybe we can go with the relay root origin directly, for now. I will check that. We need to investigate more how opengov origins get converted into `MultiLocation`s, and adjust our configuration accordingly.

Reply:
I changed the barrier here, but tests need to be updated: https://github.com/KILTprotocol/kilt-node/pull/603/commits/d9daae88e82c9693215ee1dffd6712c5a263e705.